File: backend/src/projects/controllers/TestResultController.cs
using System;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Mvc;
using Testoria.Projects.DTO;
using Testoria.Projects.Services;

namespace Testoria.Projects.Controllers
{
    /// <summary>
    /// Контроллер для управления результатами тестов
    /// </summary>
    [ApiController]
    [Route("api/[controller]")]
    public sealed class TestResultController : ControllerBase
    {
        private readonly TestResultService _testResultService;

        /// <summary>
        /// Инициализирует новый экземпляр класса TestResultController
        /// </summary>
        /// <param name="testResultService">Сервис для работы с результатами тестов</param>
        public TestResultController(TestResultService testResultService)
        {
            _testResultService = testResultService;
        }

        /// <summary>
        /// Создает результат теста
        /// </summary>
        /// <param name="createTestResultDto">Данные для создания результата теста</param>
        /// <returns>Созданный результат теста</returns>
        /// <response code="201">Результат теста успешно создан</response>
        /// <response code="400">Некорректные данные запроса</response>
        /// <response code="404">Проект не найден</response>
        [HttpPost]
        [ProducesResponseType(typeof(TestResultDto), 201)]
        [ProducesResponseType(400)]
        [ProducesResponseType(404)]
        public async Task<ActionResult<TestResultDto>> CreateTestResult([FromBody] CreateTestResultDto createTestResultDto)
        {
            if (!ModelState.IsValid)
            {
                return BadRequest(ModelState);
            }

            try
            {
                var testResult = await _testResultService.CreateTestResultAsync(createTestResultDto);
                return CreatedAtAction(nameof(GetTestResult), new { id = testResult.Id }, testResult);
            }
            catch (ProjectNotFoundException)
            {
                return NotFound($"Проект с идентификатором {createTestResultDto.ProjectId} не найден");
            }
        }

        /// <summary>
        /// Отправляет результаты тестов Jest в формате JSON
        /// </summary>
        /// <param name="projectId">Идентификатор проекта</param>
        /// <param name="jestJson">JSON данные от Jest</param>
        /// <returns>Массив созданных результатов тестов</returns>
        /// <response code="201">Результаты тестов успешно созданы</response>
        /// <response code="400">Некорректные данные запроса</response>
        /// <response code="404">Проект не найден</response>
        [HttpPost("jest/{projectId}")]
        [ProducesResponseType(typeof(TestResultDto[]), 201)]
        [ProducesResponseType(400)]
        [ProducesResponseType(404)]
        public async Task<ActionResult<TestResultDto[]>> CreateTestResultsFromJest(
            Guid projectId, 
            [FromBody] JsonElement jestJson)
        {
            try
            {
                var testResults = await _testResultService.CreateTestResultsFromJestAsync(projectId, jestJson);
                return CreatedAtAction(nameof(GetTestResultsByProject), new { projectId }, testResults);
            }
            catch (ProjectNotFoundException)
            {
                return NotFound($"Проект с идентификатором {projectId} не найден");
            }
            catch (JsonException ex)
            {
                return BadRequest($"Некорректный JSON: {ex.Message}");
            }
        }

        /// <summary>
        /// Получает результат теста по идентификатору
        /// </summary>
        /// <param name="id">Идентификатор результата теста</param>
        /// <returns>Результат теста</returns>
        /// <response code="200">Результат теста найден</response>
        /// <response code="404">Результат теста не найден</response>
        [HttpGet("{id}")]
        [ProducesResponseType(typeof(TestResultDto), 200)]
        [ProducesResponseType(404)]
        public async Task<ActionResult<TestResultDto>> GetTestResult(Guid id)
        {
            // TODO: Реализовать получение результата теста по ID
            return NotFound("Метод не реализован");
        }

        /// <summary>
        /// Получает все результаты тестов для проекта
        /// </summary>
        /// <param name="projectId">Идентификатор проекта</param>
        /// <returns>Список результатов тестов</returns>
        /// <response code="200">Список результатов тестов получен</response>
        [HttpGet("project/{projectId}")]
        [ProducesResponseType(typeof(TestResultDto[]), 200)]
        public async Task<ActionResult<TestResultDto[]>> GetTestResultsByProject(Guid projectId)
        {
            // TODO: Реализовать получение результатов тестов по проекту
            return Ok(Array.Empty<TestResultDto>());
        }
    }
}

File: backend/src/projects/controllers/test-result.controller.ts
import { 
  Controller, 
  Get, 
  Post, 
  Body, 
  Param, 
  HttpStatus, 
  HttpCode,
  NotFoundException,
  BadRequestException 
} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse, ApiParam, ApiBody } from '@nestjs/swagger';
import { TestResultService } from '../services/test-result.service';
import { CreateTestResultDto, TestResultDto, JestTestResultsDto } from '../dto/test-result.dto';
import { ProjectNotFoundException } from '../exceptions/project-not-found.exception';

/**
 * Контроллер для управления результатами тестов
 */
@ApiTags('test-results')
@Controller('test-results')
export class TestResultController {
  constructor(private readonly testResultService: TestResultService) {}

  /**
   * Создает результат теста
   * @param createTestResultDto Данные для создания результата теста
   * @returns Созданный результат теста
   */
  @Post()
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ 
    summary: 'Создание результата теста',
    description: 'Создает новый результат теста в системе'
  })
  @ApiBody({ type: CreateTestResultDto })
  @ApiResponse({ 
    status: 201, 
    description: 'Результат теста успешно создан',
    type: TestResultDto
  })
  @ApiResponse({ 
    status: 400, 
    description: 'Некорректные данные запроса'
  })
  @ApiResponse({ 
    status: 404, 
    description: 'Проект не найден'
  })
  async createTestResult(@Body() createTestResultDto: CreateTestResultDto): Promise<TestResultDto> {
    try {
      return await this.testResultService.createTestResult(createTestResultDto);
    } catch (error) {
      if (error instanceof ProjectNotFoundException) {
        throw new NotFoundException(error.message);
      }
      throw error;
    }
  }

  /**
   * Отправляет результаты тестов Jest в формате JSON
   * @param projectId Идентификатор проекта
   * @param jestData JSON данные от Jest
   * @returns Массив созданных результатов тестов
   */
  @Post('jest/:projectId')
  @HttpCode(HttpStatus.CREATED)
  @ApiOperation({ 
    summary: 'Отправка результатов Jest',
    description: 'Отправляет результаты тестов Jest в формате JSON и создает записи в базе данных'
  })
  @ApiParam({ 
    name: 'projectId', 
    description: 'Идентификатор проекта',
    example: '123e4567-e89b-12d3-a456-426614174000'
  })
  @ApiBody({ 
    description: 'JSON данные от Jest',
    schema: {
      type: 'object',
      properties: {
        testResults: {
          type: 'array',
          items: {
            type: 'object',
            properties: {
              assertionResults: {
                type: 'array',
                items: {
                  type: 'object',
                  properties: {
                    title: { type: 'string', example: 'should pass test' },
                    status: { type: 'string', enum: ['passed', 'failed', 'skipped'] },
                    duration: { type: 'number', example: 100 },
                    failureMessages: { type: 'array', items: { type: 'string' } }
                  }
                }
              }
            }
          }
        }
      }
    }
  })
  @ApiResponse({ 
    status: 201, 
    description: 'Результаты тестов успешно созданы',
    type: [TestResultDto]
  })
  @ApiResponse({ 
    status: 400, 
    description: 'Некорректные данные запроса'
  })
  @ApiResponse({ 
    status: 404, 
    description: 'Проект не найден'
  })
  async createTestResultsFromJest(
    @Param('projectId') projectId: string,
    @Body() jestData: any,
  ): Promise<TestResultDto[]> {
    try {
      return await this.testResultService.createTestResultsFromJest(projectId, jestData);
    } catch (error) {
      if (error instanceof ProjectNotFoundException) {
        throw new NotFoundException(error.message);
      }
      if (error instanceof SyntaxError) {
        throw new BadRequestException(`Некорректный JSON: ${error.message}`);
      }
      throw error;
    }
  }

  /**
   * Получает результат теста по идентификатору
   * @param id Идентификатор результата теста
   * @returns Результат теста
   */
  @Get(':id')
  @ApiOperation({ 
    summary: 'Получение результата теста по ID',
    description: 'Возвращает результат теста по его уникальному идентификатору'
  })
  @ApiParam({ 
    name: 'id', 
    description: 'Уникальный идентификатор результата теста',
    example: '123e4567-e89b-12d3-a456-426614174000'
  })
  @ApiResponse({ 
    status: 200, 
    description: 'Результат теста найден',
    type: TestResultDto
  })
  @ApiResponse({ 
    status: 404, 
    description: 'Результат теста не найден'
  })
  async getTestResult(@Param('id') id: string): Promise<TestResultDto | null> {
    return this.testResultService.getTestResultById(id);
  }

  /**
   * Получает все результаты тестов для проекта
   * @param projectId Идентификатор проекта
   * @returns Список результатов тестов
   */
  @Get('project/:projectId')
  @ApiOperation({ 
    summary: 'Получение результатов тестов проекта',
    description: 'Возвращает все результаты тестов для указанного проекта'
  })
  @ApiParam({ 
    name: 'projectId', 
    description: 'Идентификатор проекта',
    example: '123e4567-e89b-12d3-a456-426614174000'
  })
  @ApiResponse({ 
    status: 200, 
    description: 'Список результатов тестов получен успешно',
    type: [TestResultDto]
  })
  async getTestResultsByProject(@Param('projectId') projectId: string): Promise<TestResultDto[]> {
    return this.testResultService.getTestResultsByProject(projectId);
  }
}
